### PR TITLE
Add in tests, improve structure, implement courses parser

### PR DESCRIPTION
**Pull request by iheanyi**
# Description

This Pull Request will implement the fetching of courses from the API. Also, we want to add in tests to improve coverage/reliability (like any good software engineer). Some things were also moved around to keep the modules as relevant as possible to one another.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+const (
+	baseAuthTermURL = "https://ssb.cc.nd.edu/StudentRegistrationSsb/ssb/term/search?mode=search&term="
+)
+
+func setupClient() (*http.Client, error) {
+	cookieJar, err := cookiejar.New(nil)
+
+	if err != nil {
+		log.Fatal(err)
```
Comment (aybabtme) [inline client.go]: `log.Fatal` will exit the process, so `return nil, err` never happens

File: client.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	client := &http.Client{
+		Jar: cookieJar,
+	}
+
+	return client, err
+}
+
+func authenticateClient(c *http.Client, term string) {
+	authURL := baseAuthTermURL + term
+	fmt.Println(authURL)
```
Comment (aybabtme) [inline client.go]: always prefer `log.Xxx` to `fmt.Xxx`: in this case use `log.Print(authURL)` (no need to add `\n` to it, the logger does it automatically if there isn't one)

File: courses.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+func FetchAllTermCourses() (string, error) {
+	terms, err := FetchTerms()
+
+	if err != nil {
+		return "", err
+	}
+
+	for i := range terms {
```
Comment (aybabtme) [inline courses.go]: you can do:

``` go
for _, term := range terms {
    log.Print(term.Code)
}
```

**Review by aybabtme — COMMENTED** [review]
just a quick few tips

File: client.go (hunk 2 above)
Comment (iheanyi) [reply client.go]: 👍 Thank you.

File: courses.go (hunk 3 above)
Comment (iheanyi) [reply courses.go]: Oh wow, perfect! Thanks!

File: client.go (hunk 1 above)
Comment (iheanyi) [reply client.go]: 🤔 That's true. I gotta get rid of that.

File: client.go [hunk 4]
```diff
@@ -68,6 +68,7 @@ func (c *Client) NewRequest(method, urlStr string, body interface{}) (*http.Requ
 }
 
 func (c *Client) Do(req *http.Request, v interface{}) (*http.Response, error) {
+	log.Printf("url=%q", req.URL)
```
Comment (aybabtme) [inline client.go]: don't forget to remove this

File: client.go [hunk 5]
```diff
@@ -86,7 +87,10 @@ func (c *Client) Do(req *http.Request, v interface{}) (*http.Response, error) {
 				return nil, err
 			}
 		} else {
-			err := json.NewDecoder(resp.Body).Decode(v)
+			buf := bytes.NewBuffer(nil)
+			io.Copy(buf, resp.Body)
+			log.Print(buf.String())
```
Comment (aybabtme) [inline client.go]: and this :P

File: main.go [hunk 6]
```diff
@@ -28,60 +23,63 @@ func main() {
...
-	terms, err := FetchTerms()
+	c, err := NewClient(nil)
 	if err != nil {
 		log.Fatal(err)
 	}
 
-	fmt.Println(terms)
-
-	_, err = FetchTermCourses(&terms[0])
+	ts := &TermsService{client: c}
```
Comment (aybabtme) [inline main.go]: or use `ts := c.Terms`

Comment (iheanyi) [reply main.go]: Yep, definitely started refactoring towards that today. :)

**Pull request merged**